basis/msgpack/rpc/rpc.factor

```diff
@@ -0,0 +1,161 @@
+USING: msgpack kernel accessors assocs arrays sequences
+combinators make io threads destructors concurrency.locks
+io.streams.duplex math.parser ;
+
+IN: msgpack.rpc
+
+TUPLE: request msgid method params ;
+
+TUPLE: response msgid error result ;
+
+TUPLE: notification method params ;
+
+TUPLE: session active? incoming-requests outgoing-requests notification-callback request-callback ;
+
+M: request write-msgpack
+    [ msgid>> ] [ method>> ] [ params>> ]
+    tri '[ 0 , _ , _ , _ , ]
+    { } make write-msgpack ;
+
+M: response write-msgpack
+    [ msgid>> ] [ error>> ] [ result>> ]
+    tri '[ 1 , _ , _ , _ , ]
+    { } make write-msgpack ;
+
+M: notification write-msgpack
+    [ method>> ] [ params>> ]
+    bi '[ 2 , _ , _ , ]
+    { } make write-msgpack ;
+
+GENERIC: feed-packet ( session packet -- )
+GENERIC: stop ( session -- )
+
+! TODO close active requests
+M: session stop ( session -- )
+    f swap active?<< ;
+
+: <session> ( -- session )
+    t H{ } clone H{ } clone [ drop ] [ drop ] session boa ;
+
+: <request> ( msgid method params -- request )
```

Comment:
If you aren't going to do anything different, these are auto-generated using constructor syntax:

```
C: <request> request
```